### ISSUE TITLE
UseDebugLogger is not practically usable

### ISSUE THREAD

**Issue by bartelink** [EventStore/EventStore#1575]
[DebugLogger](https://github.com/EventStore/EventStore/blob/29e5d77540504d6fc5f00307b56b56ec139f5112/src/EventStore.ClientAPI/Common/Log/DebugLogger.cs#L9) uses `Debug.WriteLine`, which bears a `Conditional("DEBUG")`Attribute:-

            System.Diagnostics.Debug.WriteLine(Log("ERROR", format, args));

The net effect of this is that only a DEBUG build of the nuget package would actually ever do anything.

For me, the solution lies in
a) a change to Trace.WriteLine
b) a slight doc change to mention that it writes to the debugger output (atm it's technically not lying)

I'd be happy to send a PR to address this, but the question is what combo of naming/deprecation/breaking changes based on the following make sense:-
1) Fix/add a new impl TraceLogger/UseTraceLogger
2) remove/deprecate old impl
3) adjust docs to mention that it sends to `Trace.WriteLine` and hence will show up in debugger and via tools such as SysInternals DbgView
4) For `EventStore.Client.NetCore`, do I do the same PR or anticipate it percolating through once I put it here?

**Comment by bartelink** [comment]
Upon reading https://github.com/EventStore/EventStore/issues/1352 and seeing the (possibly-over-simplistic) simplicity of impl of https://github.com/svenschelfaut/EventStore.SerilogAdapter I personally will likely adapt the log to Serilog my side and use sinks outside of that to manage any logging requirements.

(ASIDE wrt breaking changes which nobody _actually_ wants: Arguably it would help focus what is after entitled a `ConnectionSettingsBuilder` if all of these log settings were either obsoleted (which would remove them from intellisense but not break anyone) or moved off to a separate config DSL).

Happy to have this just closed/suspended pending a conclusion being reached on the other (semantic logging) issue.

**Comment by gregoryyoung** [comment]
I would suggest adding a new provider and possibly renaming things. The Trace provider does exactly what it says it does.

Ideally we could do this by adding a SeriLogLogger. The big question is if we ship this contained or keep it as an outside provider. I do not have much issue with the SeriLog dependency as we are already including it but it may be good to ship it externally if only to show others how they can do the same.

Does this make sense?

**Comment by bartelink** [comment]
TL;DR IMO the current thing really doesn't work as anyone would expect, and adding other stuff has a cost and doesnt solve the need I think `UseDebugLogger` should address: _How does a normal user get stuff into their Debug output window in a typical usage scenario?_

> The Trace provider does exactly what it says it does.

For avoidance of doubt, I'll restate: the code uses: `Debug.WriteLine`, which, unless compiling in `DEBUG` mode, gets removed by the compiler as the method in question bears a `[Conditional("DEBUG")]`. Thus the _Debug_ provider only works in Debug builds of the client, which is not called out. I'd argue the logger should not even be present in release builds (i.e. #ifdef'd out).

Normally people would expect the feature to work in the typical use case (the NuGet). To make that work, one should:
a) switch to using `Trace.WriteLine` so it works in a release build
b) (arguably) renamed it from `DebugLogger` to remove the ambiguity (`DebugLogger` tells me `OutputDebugString`, and I think the breaking change would not be justified)

>Ideally we could do this by adding a SeriLogLogger. The big question is if we ship this contained or keep it as an outside provider. I do not have much issue with the SeriLog dependency as we are already including it but it may be good to ship it externally if only to show others how they can do the same.

I think there's a bit of a jump from 'including' to 'depending on' wrt diamond dependencies so would not do that lightly. Atm I have a wrapping layer which transforms all output to Serilog's `ILogger` with just the following (which is essentially the same as @svenschelfaut 's) so there's probably not much value aside from your point about the example:-

```
type private SerilogAdapter(log : ILogger) =
    interface EventStore.ClientAPI.ILogger with
        member __.Debug(format: string, args: obj []) =           log.Debug(format, args)
        member __.Debug(ex: exn, format: string, args: obj []) =  log.Debug(ex, format, args)
        member __.Info(format: string, args: obj []) =            log.Information(format, args)
        member __.Info(ex: exn, format: string, args: obj []) =   log.Information(ex, format, args)
        member __.Error(format: string, args: obj []) =           log.Error(format, args)
        member __.Error(ex: exn, format: string, args: obj []) =  log.Error(ex, format, args)
```

So, back to the core issue (sorry if I diluted the point with my follow-up):
- the current logger is not useful to anyone except people doing custom builds of the client (I'm actually doing that _today_ as it happens, but that ain't normal!)
- adding a third party dependency does not address that (and already has another Issue discussing progressing that)
- adding another impl with a different name `UseTraceLogger()` makes the DSL even noisier and makes for more doc work etc

**Comment by gregoryyoung** [comment]
If that is TL;DR I would hate to read the long version :)

On the other side how would deployment work? If for instance we had a dependency on a given logger should it be included? What happens when we have implementations? Might it be better to give late binding on these?

**Comment by bartelink** [comment]
(LibLog and/or Serilog etc. is obv a deep rathole; I'm OK with `ILogger` remaining as-is in general)

I'm hoping to only discuss `UseDebugLogger` in this Issue from hereon in the interests of brevity ;)
In that context, my core question boils down to:

Do you have a preference for
   a) changing the behavior of `UseDebugLogger`
   b) introducing a `UseTraceLogger`
as a way of resolving the core issue of people trying to trigger an `OutputDebugString` call using the stock `EventStore.Client` from NuGet ?
